ykrt/src/compile/j2/opt/known_bits.rs

```diff
@@ -0,0 +1,474 @@
+        assert_eq!(_iidx.index(), self.iidx_to_known_bits.len() - 1);
+    }
+}
+
+impl KnownBits {
+    /// Create an empty known bits analysis object.
+    pub(super) fn new() -> Self {
+        KnownBits {
+            iidx_to_known_bits: IndexVec::new(),
+            curr_value_pending_commit: None,
+        }
+    }
+
+    /// Returns what we known about the bits of `iidx`.
+    fn as_knownbits(&self, opt: &PassOpt, iidx: InstIdx) -> KnownBitValue {
+        if let Some(bits) = self.iidx_to_known_bits[iidx].clone() {
+            bits
+        } else {
+            KnownBitValue::unknown(match opt.inst(iidx).ty(opt) {
+                // The first two are currently unimplemented.
+                Ty::Func(_) => 0,
+                Ty::Void => 0,
+                ty => ty.bitw(),
+            })
+        }
+    }
+
+    fn set_knownbits(&mut self, bits: KnownBitValue) {
+        self.curr_value_pending_commit = Some(bits);
+    }
+
+    fn opt_and(&mut self, opt: &mut PassOpt, mut inst: And) -> OptOutcome {
+        inst.canonicalise(opt);
+        let And { tyidx, lhs, rhs } = inst;
+        let lhs_b = self.as_knownbits(opt, lhs);
+        let rhs_b = self.as_knownbits(opt, rhs);
+        let res = lhs_b.and(&rhs_b);
+        self.set_knownbits(res.clone());
+
+        // If we know the output's bits, emit that.
+        if res.all_known() {
+            return OptOutcome::Rewritten(Inst::Const(Const {
+                tyidx,
+                kind: ConstKind::Int(res.as_int()),
+            }));
+        }
+
+        // lhs = any and rhs = constant
+        // If no new information (set zeroes) is gained, that means this
+        // op was useless.
+        if rhs_b.all_known()
+            && (rhs_b
+                .zeroes()
+                .bitand(&lhs_b.known_ones().bitor(&lhs_b.unknowns))
+                .count_ones()
+                == 0)
+        {
+            return OptOutcome::Equiv(lhs);
+        }
+
+        OptOutcome::Rewritten(inst.into())
+    }
+
+    fn opt_const(&mut self, inst: Const) -> OptOutcome {
+        let Const { tyidx: _, kind } = &inst;
+        if let ConstKind::Int(kind) = kind {
+            self.set_knownbits(KnownBitValue::from_constant(kind))
+        }
+        OptOutcome::Rewritten(inst.into())
+    }
+
+    fn opt_or(&mut self, opt: &PassOpt, mut inst: Or) -> OptOutcome {
+        inst.canonicalise(opt);
+        let Or {
+            tyidx,
+            lhs,
+            rhs,
+            disjoint: _,
+        } = inst;
+        let lhs_b = self.as_knownbits(opt, lhs);
+        let rhs_b = self.as_knownbits(opt, rhs);
+        let res = lhs_b.or(&rhs_b);
+        self.set_knownbits(res.clone());
+
+        // If we know the output's bits, emit that.
+        if res.all_known() {
+            return OptOutcome::Rewritten(Inst::Const(Const {
+                tyidx,
+                kind: ConstKind::Int(res.as_int()),
+            }));
+        }
+
+        // lhs = any and rhs = constant
```

Comment:
I'm not sure this comment adds anything?

Reply:
Remove as part of documentation change in 273af03d9ea7dfbb132198e40ee07dfd31a71c5f